tests/distributions/test_pickle.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import absolute_import, division, print_function
+
+import inspect
+
+import pytest
+import six.moves.cPickle as pickle
+import torch
+
+import pyro.distributions as dist
+from pyro.distributions.torch_distribution import TorchDistributionMixin
+from tests.common import xfail_param
+
+# Collect distributions.
+BLACKLIST = [
+    dist.TorchDistribution,
+    dist.ExponentialFamily,
+    dist.OMTMultivariateNormal,
+]
+XFAIL = {
+    dist.Gumbel: xfail_param(dist.Gumbel, reason='cannot pickle weakref'),
+}
+DISTRIBUTIONS = [d for d in dist.__dict__.values()
+                 if isinstance(d, type)
+                 if issubclass(d, TorchDistributionMixin)
+                 if d not in BLACKLIST]
+DISTRIBUTIONS.sort(key=lambda d: d.__name__)
+DISTRIBUTIONS = [XFAIL.get(d, d) for d in DISTRIBUTIONS]
+
+# Provide default args if Dist(1, 1, ..., 1) is known to fail.
+ARGS = {
+    dist.AVFMultivariateNormal: [torch.zeros(3), torch.eye(3), torch.rand(2, 4, 3)],
+    dist.Bernoulli: [0.5],
+    dist.Binomial: [2, 0.5],
+    dist.Categorical: [torch.ones(2)],
+    dist.Delta: [torch.tensor(0.)],
+    dist.Dirichlet: [torch.ones(2)],
+    dist.GaussianScaleMixture: [torch.ones(2), torch.ones(3), torch.ones(3)],
+    dist.Geometric: [0.5],
+    dist.Independent: [dist.Normal(torch.zeros(2), torch.ones(2)), 1],
+    dist.LowRankMultivariateNormal: [torch.zeros(2), torch.ones(2, 2), torch.ones(2)],
+    dist.MaskedMixture: [torch.tensor([1, 0]).byte(), dist.Normal(0, 1), dist.Normal(0, 2)],
+    dist.MixtureOfDiagNormals: [torch.ones(2, 3), torch.ones(2, 3), torch.ones(2)],
+    dist.MixtureOfDiagNormalsSharedCovariance: [torch.ones(2, 3), torch.ones(3), torch.ones(2)],
+    dist.Multinomial: [2, torch.ones(2)],
+    dist.MultivariateNormal: [torch.ones(2), torch.eye(2)],
+    dist.OneHotCategorical: [torch.ones(2)],
+    dist.RelaxedBernoulli: [1.0, 0.5],
+    dist.RelaxedBernoulliStraightThrough: [1.0, 0.5],
+    dist.RelaxedOneHotCategorical: [1., torch.ones(2)],
+    dist.RelaxedOneHotCategoricalStraightThrough: [1., torch.ones(2)],
+    dist.TransformedDistribution: [dist.Normal(0, 1), torch.distributions.ExpTransform()],
+    dist.Uniform: [0, 1],
+    dist.VonMises3D: [torch.tensor([1., 0., 0.])],
+}
+
+
+@pytest.mark.parametrize('Dist', DISTRIBUTIONS)
+def test_pickle(Dist):
+    if Dist in ARGS:
+        args = ARGS[Dist]
+    else:
+        # Optimistically try to initialize with Dist(1, 1, ..., 1).
+        try:
+            # Python 3.6+
```

Comment:
should i add python 3.6 to travis? (either in addition to or in lieu of 3.5). not sure which is more common among users, but i use 3.6.2 locally.

Reply:
new builds will, which is why im thinking of replacing 3.5 since apparently (from this PR) there are breaking changes 3.5 -> 3.6, which we generally wont catch

Reply:
> should i add python 3.6 to travis? (either in addition to or in lieu of 3.5). not sure which is more common among users, but i use 3.6.2 locally.

Should we just move to testing on Python 3.6? It is not ideal, but I think a vast majority of users would either by using Python 2.7 or 3.6. Doing testing on both 3.5 and 3.6 will take up too much time, and not worth it for each commit.

Reply:
ok based on this thread it seems like everyones ok testing on 3.6. ill make the appropriate changes and submit a PR

Reply:
@neerajprad with python 3.6, we don't have to sort the keys in dict ;)

Reply:
Life would indeed be much better if we just dropped python 2 support completely! 😆 PyMC3 is doing that.